File: modules/compiler/src/test/resources/compute_benchmark_baseline.py
#!/usr/bin/env python3
#
# The Compukter Kraft Developers
#
# Copyright (C) 2026 Vsevolod Petrov (lazyhat)
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

from __future__ import annotations

import sys
import time
from collections.abc import Callable


INT_MASK = 0xFFFFFFFF
INT_SIGN = 0x80000000


def i32(value: int) -> int:
    value &= INT_MASK
    if value & INT_SIGN:
        return value - 0x100000000
    return value


def shl(value: int, bits: int) -> int:
    return i32(i32(value) << bits)


def integer_mix(iterations: int) -> int:
    state = 305_419_896
    acc = -1_640_531_527
    i = 0
    while i < iterations:
        state = i32(state * 1_664_525 + 1_013_904_223)
        x = i32(state ^ (state >> 16))
        acc = i32(i32(acc + x) ^ shl(acc, 5))
        acc = i32(acc + i32(i32(i * 31) ^ (x >> 3)))
        i += 1
    return acc


def function_mix(iterations: int) -> int:
    acc = 324_508_639
    i = 0
    while i < iterations:
        acc = function_mix_b(function_mix_a(acc, i), i)
        i += 1
    return acc


def function_mix_a(value: int, index: int) -> int:
    return i32(i32(i32(value + i32(index * 17)) ^ shl(value, 3)) + (index >> 1))


def function_mix_b(value: int, index: int) -> int:
    return i32(i32(i32(value ^ i32(index * 131)) + (value >> 2)) ^ shl(index, 4))


def branch_div(iterations: int) -> int:
    acc = 7
    i = 1
    while i < iterations + 1:
        modulo = i % 11
        if modulo == 0:
            acc = i32(acc + i // 3)
        elif modulo < 5:
            acc = i32(i32(acc ^ i32(i * 17)) + (i % 7))
        else:
            acc = i32(i32(acc - (i // (modulo + 1))) + shl(acc, 1))
        i += 1
    return acc


def recursive_fib_workload(iterations: int) -> int:
    acc = 0
    i = 0
    while i < iterations:
        n = 10 + (i % 6)
        acc = i32(acc + i32(recursive_fib(n) ^ i32(i * 31)))
        i += 1
    return acc


def recursive_fib(value: int) -> int:
    if value < 2:
        return value
    return i32(recursive_fib(value - 1) + recursive_fib(value - 2))


WORKLOADS: dict[str, Callable[[int], int]] = {
    "integer-mix": integer_mix,
    "function-mix": function_mix,
    "branch-div": branch_div,
    "recursive-fib": recursive_fib_workload,
}


def main() -> None:
    workload_name = sys.argv[1]
    iterations = int(sys.argv[2])
    samples = int(sys.argv[3])

    workload = WORKLOADS[workload_name]
    checksum = None
    best_nanos = None
    for _ in range(samples):
        started = time.perf_counter_ns()
        sample_checksum = workload(iterations)
        elapsed = time.perf_counter_ns() - started
        if checksum is None:
            checksum = sample_checksum
        elif checksum != sample_checksum:
            raise RuntimeError(f"{workload_name} checksum changed between samples: {checksum} != {sample_checksum}")
        best_nanos = elapsed if best_nanos is None else min(best_nanos, elapsed)

    print("checksum\tbest_nanos")
    print(f"{checksum}\t{best_nanos}")


if __name__ == "__main__":
    main()

File: native/ckl-vm/examples/compute_benchmark_baseline.rs
/*
 * The Compukter Kraft Developers
 *
 * Copyright (C) 2026 Vsevolod Petrov (lazyhat)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let workload = args
        .get(1)
        .unwrap_or_else(|| panic!("missing workload argument"));
    let iterations = parse_arg(&args, 2, "iterations");
    let samples = parse_arg(&args, 3, "samples");
    let warmup_iterations = parse_arg(&args, 4, "warmup_iterations");

    if warmup_iterations > 0 {
        run_workload(workload, warmup_iterations);
    }

    let mut checksum = None;
    let mut best_nanos = u128::MAX;
    for _ in 0..samples {
        let started = Instant::now();
        let sample_checksum = run_workload(workload, iterations);
        let elapsed = started.elapsed().as_nanos();
        if let Some(expected) = checksum {
            assert_eq!(
                expected, sample_checksum,
                "{workload} native benchmark checksum changed between samples"
            );
        } else {
            checksum = Some(sample_checksum);
        }
        best_nanos = best_nanos.min(elapsed);
    }

    println!("checksum\tbest_nanos");
    println!(
        "{}\t{}",
        checksum.expect("at least one sample is required"),
        best_nanos
    );
}

fn parse_arg(args: &[String], index: usize, name: &str) -> i32 {
    args.get(index)
        .unwrap_or_else(|| panic!("missing {name} argument"))
        .parse()
        .unwrap_or_else(|error| panic!("invalid {name} argument: {error}"))
}

fn run_workload(workload: &str, iterations: i32) -> i32 {
    match workload {
        "integer-mix" => integer_mix(iterations),
        "function-mix" => function_mix(iterations),
        "branch-div" => branch_div(iterations),
        "recursive-fib" => recursive_fib_workload(iterations),
        _ => panic!("unknown workload: {workload}"),
    }
}

fn integer_mix(iterations: i32) -> i32 {
    assert!(iterations >= 0, "iterations must be non-negative");
    let mut state: i32 = 305_419_896;
    let mut acc: i32 = -1_640_531_527;
    let mut i: i32 = 0;
    while i < iterations {
        state = state.wrapping_mul(1_664_525).wrapping_add(1_013_904_223);
        let x = state ^ (state >> 16);
        let previous_acc = acc;
        acc = previous_acc.wrapping_add(x) ^ previous_acc.wrapping_shl(5);
        acc = acc.wrapping_add(i.wrapping_mul(31) ^ (x >> 3));
        i += 1;
    }
    acc
}

fn function_mix(iterations: i32) -> i32 {
    assert!(iterations >= 0, "iterations must be non-negative");
    let mut acc: i32 = 324_508_639;
    let mut i: i32 = 0;
    while i < iterations {
        acc = function_mix_b(function_mix_a(acc, i), i);
        i += 1;
    }
    acc
}

fn function_mix_a(value: i32, index: i32) -> i32 {
    (value.wrapping_add(index.wrapping_mul(17)) ^ value.wrapping_shl(3)).wrapping_add(index >> 1)
}

fn function_mix_b(value: i32, index: i32) -> i32 {
    (value ^ index.wrapping_mul(131)).wrapping_add(value >> 2) ^ index.wrapping_shl(4)
}

fn branch_div(iterations: i32) -> i32 {
    assert!(iterations >= 0, "iterations must be non-negative");
    let mut acc: i32 = 7;
    let mut i: i32 = 1;
    while i < iterations + 1 {
        let modulo = i % 11;
        acc = if modulo == 0 {
            acc.wrapping_add(i / 3)
        } else if modulo < 5 {
            (acc ^ i.wrapping_mul(17)).wrapping_add(i % 7)
        } else {
            acc.wrapping_sub(i / (modulo + 1))
                .wrapping_add(acc.wrapping_shl(1))
        };
        i += 1;
    }
    acc
}

fn recursive_fib_workload(iterations: i32) -> i32 {
    assert!(iterations >= 0, "iterations must be non-negative");
    let mut acc: i32 = 0;
    let mut i: i32 = 0;
    while i < iterations {
        let n = 10 + (i % 6);
        acc = acc.wrapping_add(recursive_fib(n) ^ i.wrapping_mul(31));
        i += 1;
    }
    acc
}

fn recursive_fib(value: i32) -> i32 {
    if value < 2 {
        value
    } else {
        recursive_fib(value - 1).wrapping_add(recursive_fib(value - 2))
    }
}
